src/libstore/realisation.hh

```diff
@@ -32,8 +54,60 @@ struct Realisation {
 
     nlohmann::json toJSON() const;
     static Realisation fromJSON(const nlohmann::json& json, const std::string& whence);
+
+    StorePath getPath() const { return outPath; }
+
+    GENERATE_CMP(Realisation, me->id, me->outPath);
+};
+
+struct OpaquePath {
+    StorePath path;
+
+    StorePath getPath() const { return path; }
+
+    GENERATE_CMP(OpaquePath, me->path);
 };
 
-typedef std::map<DrvOutput, Realisation> DrvOutputs;
+
+/**
+ * A store path with all the history of how it went into the store
+ */
+struct RealisedPath {
+    /*
+     * A path is either the result of the realisation of a derivation or
+     * an opaque blob that has been directly added to the store
+     */
+    using Raw = std::variant<Realisation, OpaquePath>;
+    Raw raw;
+
+    using Set = std::set<RealisedPath>;
+
+    RealisedPath(StorePath path) : raw(OpaquePath{path}) {}
+    RealisedPath(Realisation r) : raw(r) {}
+
+    /**
+     * Syntactic sugar to run `std::visit` on the raw value:
+     * path.visit(blah) == std::visit(blah, path.raw)
```

Comment:
TBH I wouldn't mind getting rid of the use of `std::visit`, which often seems unnecessarily verbose/complicated because of the use of closures compared to
```
if (auto realisation = variant.get_if<Realisation>)
  do something 
else if (auto realisation = variant.get_if<OpaquePath>)
  do something else
```

Reply:
I don't know in the general case (it's indeed very verbose and easily hard to read, but at the same time the exhaustiveness check is a killer feature…). But in this precise case, there was only one single call to `visit`, so it's definitely not worth it indeed.

Fixed in 4f84b2c4d638ace8e745fb78860a549217dbc00d

Reply:
@regnat For some reason that commit isn't showing up in this PR, did you push it?